os/fs/smartfs/smartfs_smart.c

```diff
@@ -1996,20 +2019,20 @@ int smartfs_rename(struct inode *mountpt, const char *oldrelpath, const char *ne
 		}
 
 		/* Okay, we will create newfilename in the newentry directory */
-		newparentdirsector = newentry.firstsector;
+		newentry.dsector = newentry.firstsector;
```

Comment:
this is incorrect, I think..

Reply:
1. finddirentry returns newentry as the entry of the parent directory.
2. So newentry.firstsector is actually the 1st data sector of new parent directory.
3. Now, all sectors in this directory are scanned to find if the entry already exists.
4. If does not exist,
     dsector = firstsector and then find available entry in one of the directory sectors to write new entry
5. dsector and last_dsector get updated ahead. firstsector too.

Reply:
New file must be not exist, if there is new file exist, it should return -EEXIST. but now it overwrite.
Current this will move oldpath to subdirectory of newpath, because dsector will be changed to firstsector.
For example,
```
mkdir("/mnt/old", 0777);
mkdir("/mnt/old/a", 0777);
mkdir("/mnt/old/b", 0777);
mkdir("/mnt/new", 0777);
mkdir("/mnt/new/c", 0777);
mkdir("/mnt/new/d", 0777);
rename("/mnt/old", "/mnt/new");
```
Result of this test will shows /mnt/new has 3 directories,  'a', 'b', and 'old'

Reply:
Ohkkay, now I got your concern regarding rename.
Actually I made an assumption earlier, I thought rename is only intended for files and it was made that way only.
From the 1st time I looked at smartfs_rename() function, I knew this would only work for renaming files.

This function WILL NOT work for renaming directories. I think it was not written that way from the very beginning.